src/query/src/within_filter.rs

```diff
@@ -0,0 +1,359 @@
+/// `WITHIN` syntax.
+///
+/// # How It Works
+/// 1. The rule analyzer detects `within(ts, timestamp)` functions in a `LogicalPlan`.
+/// 2. It infers the precision of the given timestamp (year, month, day, second, etc.).
+/// 3. Based on that precision, the rule calculates the appropriate start and end timestamps (e.g., `start = 2025-04-19 23:50:00`, `end = 2025-04-19 23:51:00`).
+/// 4. The `within` function is converted to a range expression like `ts >= start AND ts < end`.
+///
+/// # Examples
+/// - convert `ts WITHIN '2025'`
+///   to:
+///   ```sql
+///   ts >= '2025-01-01 00:00:00' AND ts < '2026-01-01 00:00:00'
+///   ```
+/// - convert `ts WITHIN '2025-04-19'`
+///   to:
+///   ```sql
+///   ts >= '2025-04-19 00:00:00' AND ts < '2025-04-20 00:00:00'
+///   ```
+/// - convert WITHIN '2025-04-19 23:50'`
+///   to:
+///   ```sql
+///   ts >= '2025-04-19 23:50:00' AND ts < '2025-04-19 23:51:00'
+///   ```
+pub struct WithinFilterRule {}
+
+impl WithinFilterRule {
+    pub fn new() -> Self {
+        WithinFilterRule {}
+    }
+}
+
+impl ExtensionAnalyzerRule for WithinFilterRule {
+    fn analyze(
+        &self,
+        plan: LogicalPlan,
+        _ctx: &QueryEngineContext,
+        _config: &ConfigOptions,
+    ) -> DFResult<LogicalPlan> {
+        plan.transform(|plan| match plan.clone() {
+            LogicalPlan::Filter(filter) => {
+                if let Expr::ScalarFunction(func) = &filter.predicate
+                    && func.func.name() == WITHIN_FILTER_NAME
+                {
+                    ensure!(
+                        func.args.len() == 2,
+                        WithinFilterInternalSnafu {
+                            message: "expected 2 arguments",
+                        }
+                    );
+                    let column_name = func.args[0].clone();
+                    let time_arg = func.args[1].clone();
+                    if let Expr::Literal(literal) = time_arg
+                        && let ScalarValue::Utf8(Some(s)) = literal
+                    {
+                        if let Some((start, end)) = try_to_infer_time_range(&s) {
+                            return Ok(Transformed::yes(convert_plan(
+                                filter.input,
+                                &column_name,
+                                start,
+                                end,
+                            )?));
+                        }
+                    }
+                    Err(DataFusionError::Plan(
+                        "Failed to convert within filter to normal filter.".to_string(),
+                    ))
+                } else {
+                    Ok(Transformed::no(plan))
+                }
+            }
+            _ => Ok(Transformed::no(plan)),
+        })
+        .map(|t| t.data)
+    }
+}
+
+/// Infers the time range from a given timestamp string.
+fn try_to_infer_time_range(timestamp: &str) -> Option<(Timestamp, Timestamp)> {
```

Comment:
Good job! I think we must take care of the timezone. We can retrieve the timezone from `QueryEngineContext -> QueryEngineContext -> pub fn timezone(&self) -> Timezone`. The `QueryEngineContext` is already in `analyze` signature.

I found that the NativeDataTime has a function to change timezone

https://docs.rs/chrono/latest/chrono/naive/struct.NaiveDateTime.html#method.and_local_timezone